File: lib/compiler/categorizer.dart
import 'package:bliss/compiler/errors.dart';
import 'package:bliss/compiler/node.dart';

class Categorizer {
  final List<Node> _stack = [];
  Node _current = ListNode([]);

  Node _toNode(String token) {
    if (_current is StringNode) {
      return StringNode(token);
    }

    return switch (num.tryParse(token)) {
      null => SymbolNode(token),
      final number => NumberNode(number),
    };
  }

  void _categorizeOpening(Node newCurrent) {
    _stack.add(_current);
    _current = newCurrent;
  }

  void _categorizeClosing({required String stringRepr}) {
    try {
      final newCurrent = _stack.removeLast();
      newCurrent.add(_current);
      _current = newCurrent;
    } on StateError {
      throw UnexpectedNodeError("Unexpected node: $stringRepr");
    }
  }

  void _categorizeElse(String token) {
    _current.add(_toNode(token));
  }

  List<String> _categorizeComment(List<String> tokens) {
    return tokens.skipWhile((x) => x != ";").skip(1).toList();
  }

  Node categorize(List<String> inTokens) {
    var tokens = inTokens;

    while (tokens.isNotEmpty) {
      final token = tokens.removeAt(0);

      // print(
      //   "TOKENS: $tokens\nTOKEN: $token\nSTACK: $_stack\nCURRENT: $_current\n\n",
      // );

      switch (token) {
        case ";":
          tokens = _categorizeComment(tokens);

        case ")":
          _categorizeClosing(stringRepr: ")");
        case "]":
          _categorizeClosing(stringRepr: "]");
        case "\""
            when token is StringNode &&
                (_current.contents as String).endsWith("\\"):
          _categorizeClosing(stringRepr: "\"");

        case "(":
          _categorizeOpening(ListNode([]));
        case "[":
          _categorizeOpening(ValueListNode([]));
        case "\"":
          _categorizeOpening(StringNode(""));

        default:
          _categorizeElse(token);
      }
    }

    return _current;
  }
}

File: src/Compiler/Categorizer.hs
module Compiler.Categorizer (categorize) where

import Compiler.Mod (CompilerError (UnexpectedNodeError), Node (ListNode, NumberNode, StringNode, SymbolNode, ValueListNode), addToNode)
import Data.List (unsnoc)
import Text.Read (readMaybe)
import Utils (throw)

data State = State
  { tokens :: [String],
    stack :: [Node],
    current :: Node
  }
  deriving (Show)

toNode :: State -> String -> Node
toNode _state token = case readMaybe token of
  Just number -> NumberNode number
  Nothing -> SymbolNode token

categorizeString :: State -> State
categorizeString state = case tokens state of
  stringContent : "\"" : newTokens ->
    let newCurrent = addToNode (current state) $ StringNode stringContent
     in state {tokens = newTokens, current = newCurrent}
  _ -> throw $ UnexpectedNodeError "Unexpected node in string"

categorizeComment :: State -> State
categorizeComment state =
  let newTokens = drop 1 $ dropWhile (/= ";") $ tokens state
   in state {tokens = newTokens}

categorizeClosing :: State -> String -> State
categorizeClosing state strRep = case unsnoc $ stack state of
  Just (newStack, newCurrent) ->
    let modifiedNewCurrent = addToNode newCurrent $ current state
     in state {stack = newStack, current = modifiedNewCurrent}
  Nothing -> throw $ UnexpectedNodeError $ "Unexpected node: " ++ strRep

categorizeOpening :: State -> Node -> State
categorizeOpening state newCurrent =
  let newStack = stack state ++ [current state]
   in state {stack = newStack, current = newCurrent}

categorizeElse :: State -> String -> State
categorizeElse state token =
  let newCurrent = addToNode (current state) $ toNode state token
   in state {current = newCurrent}

categorizeToken :: State -> String -> State
categorizeToken state "\"" = categorizeString state
categorizeToken state ";" = categorizeComment state
categorizeToken state ")" = categorizeClosing state ")"
categorizeToken state "]" = categorizeClosing state "]"
categorizeToken state "(" = categorizeOpening state $ ListNode []
categorizeToken state "[" = categorizeOpening state $ ValueListNode []
categorizeToken state token = categorizeElse state token

categorize :: [String] -> Node
categorize inTokens =
  let aux state = case tokens state of
        [] -> current state
        token : rest ->
          let restState = state {tokens = rest}
              categorizedState = categorizeToken restState token
              newState = categorizedState {tokens = tokens categorizedState}
           in aux newState
      initialState = State {tokens = inTokens, stack = [], current = ListNode []}
   in aux initialState